### DISCUSSION TITLE
how to use typchain Contract type defs in useContract[Event,Read,Write] hooks?

### DISCUSSION THREAD

**Discussion by the-vampiire** [wevm/wagmi#518] [Q&A] [answered]
first off thanks for the killer lib. i found this after hand-rolling so many hooks that you guys already implemented! working on refactoring to use wagmi and ran into an issue.

i was trying to use `useContractEvent` and would like to have the event names / handler args picked up from my contract typedefs (generated by typechain).

i tried the following:

```ts
import { useContractEvent } from "wagmi";
import { MyContract } from "@evm/types"; // monorepo alias to evm dir, this points to the typechain dir in there

// Type 'MyContract' does not satisfy the constraint 'Contract'.
useContractEvent<MyContract>(...)
```

> ~~full error~~ this was due to my monorepo settings, ignore

```sh
# ignore (see note above)
```

**Comment by the-vampiire** [comment]
~~this is a weird error. thinking it could be related to the typescript alias for the import. so i tried copying over the types into the `client/` dir (where `useContractEvent` is being used).

this solved the type error, but if anyone knows how to fix it without manually copying over that would be awesome.~~

it does not however provide any suggestions or type safety for the event names and handler args. 

anyone know if this is possible?

```ts
wagmi.useContractEvent<MyContract>(
  {
    addressOrName: MyContractDeployment.contractAddress,
    contractInterface: MyContract.abi,
  },
  "<event | filter>", // autocomplete ?
  (some: string, made: number, up: string, args: number) => {}, // autocomplete for params?
);
```

**Comment by aalmada** [comment]
I'm also using typechain and I'm now trying to use wagmi. 
I have multiple event handlers but now they show this error on the console: `Uncaught TypeError: this.provider.on is not a function`. 
Do I have to use `useContractEvent`? 
The thing I don't about this option is that typechain wraps all this complexity with type checking.

**Reply by the-vampiire** [reply]
so i found this strange as well. its unclear from the docs but **you have to set the provider / signer when creating the contract** with `useContract`. **it does not get picked up automatically for you**. this seems silly because a contract with no provider/signer is effectively useless (it can neither read nor write).

in the case of `useContractEvent` **it does set the provider automatically** and will first check if there is a websocket provider and if not then fall back to the http provider.

**Comment by twopats** [comment]
I too am looking for any discussions further on type definitions for contracts or typechain friendly integration =)

**Comment by Yuripetusko** [comment]
Yeah this is one thing that useDapp handled really nicely and is missing here in wagmi

**Comment by martindavid** [comment]
Anyone has a good example on how to make the integration works nicely? I ended up get rid of typechain as I still haven't found a way to make both works.

**Comment by tmm** [comment, marked as answer]
Type-safety is coming to wagmi soon in https://github.com/wagmi-dev/wagmi/pull/941! No TypeChain required. wagmi will infer types directly based on ABIs (powered by [ABIType](https://github.com/wagmi-dev/abitype)).
